apps/arena/lib/arena/game_updater.ex

```diff
@@ -529,52 +535,55 @@ defmodule Arena.GameUpdater do
       collided_entity =
         decide_collided_entity(projectile, collides_with, external_wall_id, players_acc)
 
-      # #247 Refactor projectile collision resolution
-      case {
-        collided_entity,
+      apply_collision_updates(
+        projectile,
         Map.get(players, collided_entity),
-        Map.get(obstacles, collided_entity)
-      } do
-        # Projectile hit nothing
-        {nil, nil, nil} ->
-          {projectiles_acc, players_acc}
-
-        # Projectile hit the external wall
-        {entity_id, _, _} when entity_id == external_wall_id ->
-          projectile = put_in(projectile, [:aditional_info, :status], :EXPLODED)
-          {Map.put(projectiles_acc, projectile_id, projectile), players_acc}
-
-        # Projectile hit a player
-        {_entity_id, player, nil} ->
-          player = Player.change_health(player, projectile.aditional_info.damage)
-
-          projectile = put_in(projectile, [:aditional_info, :status], :EXPLODED)
-
-          if player.aditional_info.health == 0 do
-            send(self(), {:to_killfeed, projectile.aditional_info.owner_id, player.id})
-          end
-
-          {
-            Map.put(projectiles_acc, projectile_id, projectile),
-            Map.put(players_acc, player.id, player)
-          }
-
-        # Projectile hit an obstacle
-        {_entity_id, nil, obstacle} when not is_nil(obstacle) ->
-          projectile = put_in(projectile, [:aditional_info, :status], :EXPLODED)
-
-          {
-            Map.put(projectiles_acc, projectile_id, projectile),
-            players_acc
-          }
-
-        # Projectile hit something else
-        _ ->
-          {projectiles_acc, players_acc}
-      end
+        Map.get(obstacles, collided_entity),
+        collided_entity == external_wall_id,
+        accs
+      )
     end)
   end
 
+  defp apply_collision_updates(projectile, player, obstacle, collided_with_external_wall?, accs)
+
+  # Projectile collided an obstacle or went outside of the arena
+  defp apply_collision_updates(
+         projectile,
+         nil,
+         obstacle,
+         collided_with_external_wall?,
+         {projectiles_acc, players_acc}
+       )
+       when not is_nil(obstacle) or collided_with_external_wall? do
+    projectile = put_in(projectile, [:aditional_info, :status], :EXPLODED)
+
+    {
+      Map.put(projectiles_acc, projectile.id, projectile),
+      players_acc
+    }
+  end
+
+  # Projectile collided a player
+  defp apply_collision_updates(projectile, player, _, _, {projectiles_acc, players_acc})
+       when not is_nil(player) do
```

Comment:
Technically redundant this `when not is_nil(player)` due to the above clause

Reply:
With this we're checking that the projectile actually collided with a player, if we remove it this would be the default clause and we don't want that, the nil match is prioritizing the player collisions with over the obstacles/wall ones

Reply:
You are right